invenio_sipstore/models.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2016 CERN.
+#
+# Invenio is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Invenio is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Invenio-SIPStore database models."""
+
+from __future__ import absolute_import, print_function
+from sqlalchemy_utils.models import Timestamp
+from sqlalchemy_utils.types import JSONType, UUIDType
+from invenio_files_rest.models import FileInstance
+from invenio_accounts.models import User
+from invenio_pidstore.models import PersistentIdentifier
+
+from invenio_db import db
+import uuid
+
+
+class SIP(db.Model, Timestamp):
+    """Submission Information Package model."""
+
+    __tablename__ = 'sipstore_sip'
+
+    id = db.Column(UUIDType, primary_key=True, default=uuid.uuid4)
+    """Id of SIP."""
+
+    sip_format = db.Column(db.String(7), nullable=False)
+    """Format of the SIP content ('json', 'marcxml' etc.)"""
+
+    content = db.Column(db.Text, nullable=False)
+    """Text blob of the SIP content."""
+
+    user_id = db.Column(db.Integer,
+                        db.ForeignKey(User.id),
+                        nullable=True,
+                        default=None)
+    """User responsible for the SIP."""
+
+    agent = db.Column(JSONType, default=lambda: dict(), nullable=False)
+    """Agent information regarding given SIP."""
+
+    #
+    # Relationships
+    #
+    user = db.relationship(User, backref='sips', foreign_keys=[user_id])
+    """Relation to the User responsible for the SIP."""
+
+
+class SIPFile(db.Model, Timestamp):
+    """Extra SIP info regarding files."""
+
+    __tablename__ = 'sipstore_sipfile'
+
+    sip_id = db.Column(UUIDType, db.ForeignKey(SIP.id))
+    """Id of SIP."""
+
+    filename = db.Column(db.String(255), nullable=False)
+    """Filename of submitted file within the SIP record."""
+
+    file_id = db.Column(UUIDType, db.ForeignKey(FileInstance.id),
+                        primary_key=True,
+                        nullable=False)
```

Comment:
We should probably add ``ondelete='RESTRICT'`` to prevent the file instance from being deleted.